src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs

```diff
@@ -22,30 +23,45 @@ namespace Microsoft.CodeAnalysis.LanguageServer.Handler.SemanticTokens
     internal class SemanticTokensHelpers
     {
         /// <summary>
-        /// Returns the semantic tokens data for a given document with an optional range.
+        /// Returns the semantic tokens data for a given document with an optional ranges.
         /// </summary>
         public static async Task<int[]> ComputeSemanticTokensDataAsync(
             ClientCapabilities capabilities,
             Document document,
-            LSP.Range? range,
+            LSP.Range[]? ranges,
             ClassificationOptions options,
             CancellationToken cancellationToken)
         {
-            var tokenTypesToIndex = SemanticTokensSchema.GetSchema(capabilities.HasVisualStudioLspCapability()).TokenTypeToIndex;
+            if (ranges is not null)
+            {
+                // Expecting sorted ranges within the document
+                Array.Sort(ranges, static (r1, r2) => r1.CompareTo(r2));
+            }
 
+            var tokenTypesToIndex = SemanticTokensSchema.GetSchema(capabilities.HasVisualStudioLspCapability()).TokenTypeToIndex;
             var root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
             var text = await document.GetValueTextAsync(cancellationToken).ConfigureAwait(false);
+            var classifiedSpansLength = ranges?.Length ?? 1;
+            using var _ = ArrayBuilder<ClassifiedSpan>.GetInstance(classifiedSpansLength, out var classifiedSpans);
 
-            // By default we calculate the tokens for the full document span, although the user 
-            // can pass in a range if they wish.
-            var textSpan = range is null ? root.FullSpan : ProtocolConversions.RangeToTextSpan(range, text);
+            for (var i = 0; i < classifiedSpansLength; i++)
+            {
+                // By default we calculate the tokens for the full document span, although the user 
+                // can pass in a range if they wish.
+                var textSpan = ranges is null ? root.FullSpan : ProtocolConversions.RangeToTextSpan(ranges[i], text);
 
-            var classifiedSpans = await GetClassifiedSpansForDocumentAsync(
-                document, textSpan, options, cancellationToken).ConfigureAwait(false);
+                var spansForRange = await GetClassifiedSpansForDocumentAsync(
+                    document, textSpan, options, cancellationToken).ConfigureAwait(false);
+
+                if (spansForRange is not null && spansForRange.Length != 0)
```

Comment:
>spansForRange.Length != 0 [](http://example.com/codeflow?start=49&length=25)

nit: I'm guessing this is rare enough that you might just want to consider removing this part of the check.